### ISSUE TITLE
Investigate using authenticated AJAX calls

### ISSUE THREAD

**Issue by mtorres253** [18F/micropurchase#708]
So that we can continue to improve the user experience for vendors and admins, we should think about implementing authenticated AJAX calls so that users don't have to navigate through so many pages to do a thing (like bid)

**Comment by harrisj** [comment]
I think this is excellent, provided this will also be accessible (I don't know enough about AJAX and accessibility to be able to do this myself). Since we use a session cookie for authentication, all AJAX requests should convey the same cookie as any other browser requests though.

**Comment by adelevie** [comment]
If there was a way for the user to do the oauth dance client-side, we could simply use the GitHub token to access the Micropurchase API.

**Comment by adelevie** [comment]
Another idea, let me know if this is bad/unsafe:

When the user logs in (using server-side OAuth), we could stick the access token inside the DOM, so it'll be accessible to client-side JS.

I think this Railscast describes what would work here:

http://railscasts.com/episodes/324-passing-data-to-javascript?view=asciicast

So something like this:

```
<%= javascript_tag do %>
  window.githubAccessToken = '<%= @github_access_token %>';
  $('a place-bid').onClick(function() {
    // add window.githubAccessToken into data variable
    $.ajax({
      type: "POST",
      url: url,
      data: data,
      success: success,
      dataType: dataType
    });
  });
<% end %>
```

**Comment by harrisj** [comment]
I don't think we need to do any of this. AJAX is just like any regular browser access and would be submitted with the session cookie that is provided with any other web page request, right?

**Comment by adelevie** [comment]
> AJAX is just like any regular browser access

I guess it could be ... what would that look like?

As I understand it, AJAX implies there's an API call. We already have an API, which is authenticated using the GitHub token. Are you saying we could/should make the API session authenticatable as well?

**Comment by harrisj** [comment]
Ah, I see what our confusion is. You're indeed right that AJAX requests would want to get a data response vs. raw HTML. So, this does seem like it would be an API request, but it would be posted from the browser with the session cookies. So, I think to support this, the AJAX would call the API but we would allow the API to authenticate with a session cookie as well.

**Comment by nickbristow** [comment]
@mtorres253 ajax is fine for accessibility. Needs review like anything else, but nothing inherently wrong with it.
